### DISCUSSION TITLE
Sleep pattern returns server error (500)

### DISCUSSION THREAD

**Discussion by voss** [babybuddy/babybuddy#349] [Support] [answered]
Hi baby buddies,

I'm running Baby Buddy on a Linode "Nanode" instance, and it's been working flawlessly since April. I've installed it following the instructions in the install guide from github, for a manual deployment.

The Nanode is running Debian 10 (Buster), and has, as mentioned worked flawlessly since April when I installed it.

I haven't touched the config of either Baby Buddy or Debian in that timeframe, but I'm now no longer able to load the sleep pattern. When loading example.com/children/child/reports/sleep/pattern I get a white page that says:

**Server error (500)**

I checked the nginx error_log in `/var/log/nginx` but that is empty (which I realise may be because I haven't configured nginx properly).

I took a look at `/var/log/uwsgi/app/babybuddy.log` for any app errors, and I see entries like this, when I get the above mentioned error page:

`[pid: 11720|app: 0|req: 39459/65061] 94.131.131.16 () {50 vars in 1100 bytes} [Fri Nov 19 11:16:27 2021] GET /children/child/reports/sleep/pattern/ => generated 145 bytes in 258 msecs (HTTP/1.1 500) 7 headers in 223 bytes (1 switches on core 0)`

Am I looking in the right place for any errors, or do Baby Buddy log errors somewhere else as well?

**Comment by cdubz** [comment]
The easiest way to get detailed error information is to enable debug mode and try to the load graph. You can do by setting a `DEBUG` environment variable to `True` (from an `.env` file if you're using one or just at CLI) or just setting the value to `True` in [settings/base.py#L25](https://github.com/babybuddy/babybuddy/blob/master/babybuddy/settings/base.py#L25). Restart the services after that change and then try to load the graph again -- you should get much more detailed error output.

What version of Baby Buddy are you running? There have been a few bug fixes related to graphs recently so you may also want to try updating to the latest version as well.

**Comment by voss** [comment]
Thank you for your **very** fast reply!

I'm running version: v1.5.1 (4df001c). Good shout on potentially updating to the latest version. Can I go straight to the latest, 1.9.1 release, or should I update in steps to some of the previous updates first? And also, on that, is there a guide on how not to nuke your install when updating manually? (:

I will change DEBUG to True in the base.py file, and see what error output I get. Will the error output be sent directly to the error page, or should I still look in the babybuddy.log file?

Thanks a lot!

**Comment by cdubz** [comment]
You _should_ be able to up the latest version fine, just make sure to run migrations after updating. And definitely make a database backup beforehand just in case.

Debug output will go to browser.

**Comment by voss** [comment]
Thanks for your help @cdubz.

I tried `git pull`, and was told that `settings/base.py` and `urls.py` had local changes, and had to be stashed or committed. I stashed them, ran `git pull`, and then `git stash pop` to merge those locally changed files. I guess merging that 1.5.1 `base.py` with the one from 1.9.1 would screw things up. I don't remember changing anything in `urls.py` though. After running `git stash pop` I then tried to run migrate which gave me this error:

```
/var/www/babybuddy/public:> python manage.py migrate
Traceback (most recent call last):
  File "manage.py", line 4, in <module>
    from dotenv import load_dotenv, find_dotenv
ImportError: No module named dotenv
```

After restarting uwsgi and nginx, I was met with a white page saying "internal server error" when loading, so I quickly rolled back as I was under some pressure to make this excellent app work again (:

I'm sorry for being so helpless here, but do you know if there's something else I should have done before running `python manage.py migrate`, or would this come down to me stashing the old base.py file?

After rolling back to 1.5.1, I enabled the debugging mode, and when trying to view the sleep pattern chart, I now get a much more verbatim error message which has this at the core:

```
Exception Type: TypeError
Exception Value: object of type 'float' has no len()
Exception location: ./reports/graphs/sleep_pattern.py, line 100, in sleep_pattern
```

I will try doing another upgrade later in the week where I don't stash anything, that might fix the error message I got from running migrate. Upgrading to 1.9.1 definitely seems like an easier solution, if I can get the update to run properly.

**Comment by cdubz** [comment]
Ah you pretty far back so you'll need to re-run `pipenv install --three` as well to make sure you have all the latest python dependencies. _Then_ running migrations should work (side note -- you should tun the migrations through `pipenv` so `pipenv run python manage.py migrate` to ensure you use the environment packages.

If you can share the diffs for the settings and urls files I can possible also help you get that a bit better sorted, if necessary.

**Comment by voss** [comment]
Perfect. I think I will just try the update, now that I understand the process a bit better.

I was thinking of making a copy of the `/var/www/babybuddy` folder, and then update the app for that folder, to see if I can do it without any errors. My question is, updating a copy of the app, in that folder, wouldn't affect the instance currently running, as I'm only updating the virtual environment for that copy. Is that correct?

Sorry if these are very stupid questions, I have very little experience with pip and virtual environments (:

**Comment by cdubz** [comment, marked as answer]
No worries. Python's package management system is kind of shitshow so there are no stupid questions (:

Copying the folder as a backup _should_ work because it should have the `.git` and `.venv` (where virtual environment libraries live) folders. That said, I'd recommend just creating a backup of the instance that you can fully restore from if something goes wrong (assuming that's possible with Linode -- I have never used it). Also, in theory, if something did get hosed you should still be able to checkout v1.4.1 and re-run `pipenv install --three` to get the correct packages back in place.

Also you may need to want to `export PIPENV_VENV_IN_PROJECT=1` before any `pipenv install` commands. I don't _think_ its necessary after the first time but I'm not 100% sure on that.

So my recommendation here is:

1. Backup Linode instance or `tar -cxf ~/babybuddy.tar.gz /var/www/babybuddy`
2. `cd /var/www/babybuddy`
3. `git add --all && git stash`
4. `git fetch && git checkout v1.9.1`
5. `git stash pop`
6. `export PIPENV_VENV_IN_PROJECT=1` (just in case)
7. `pipenv install --three`
8. `pipenv python manage.py migrate`
9. `sudo service uwsgi restart`

And if you run in to irreparable issues anywhere along the way -- just restore the Linode instance backup or unpack the directory contents.

**Reply by voss** [reply]
Thank you so much for your patience and help @cdubz, this worked a treat!

Updated to 1.9.1. and now have sleep pattern again, and all the other new shiny things (:
